### PR TITLE
[SYCL] Add support for RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR macro

### PR DESCRIPTION
**Pull request by Fznamznon**
This macro allows to enable diagnostics of writing to raw pointers
obtained from constant_ptr.

Signed-off-by: Mariya Podchishchaeva <mariya.podchishchaeva@intel.com>

### CODE REVIEW

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Overview
+
+This file describes macroses that are having effect on SYCL compiler and run-time.
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
This file describes macros that are having effect on SYCL compiler and run-time.
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 2]
```diff
@@ -669,9 +678,18 @@ class accessor :
   using AccessorSubscript =
       typename AccessorCommonT::template AccessorSubscript<Dims>;
 
-  using RefType = DataT &;
   using ConcreteASPtrType = typename detail::PtrValueType<DataT, AS>::type *;
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
+  using RefType = typename std::conditional<
+      AS == access::address_space::constant_space,
+      const DataT &, DataT &>::type;
```
Comment (AGindinson) [inline sycl/include/CL/sycl/accessor.hpp]: It's kinda funny that clang-format chose to treat identical blocks of code differently. Could this be aligned with the other `RefType` declarations?

File: sycl/include/CL/sycl/multi_ptr.hpp [hunk 3]
```diff
@@ -70,9 +77,21 @@ template <typename ElementType, access::address_space Space> class multi_ptr {
     return *this;
   }
 
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
+  using ReturnPtr =
+      typename std::conditional<Space == access::address_space::constant_space,
+                                const ElementType *, ElementType *>::type;
+  using ReturnRef =
```
Comment (AGindinson) [inline sycl/include/CL/sycl/multi_ptr.hpp]:
```suggestion
  using ReturnRef = ReturnConstRef
```

File: sycl/include/CL/sycl/multi_ptr.hpp [hunk 4]
```diff
@@ -358,11 +382,18 @@ template <access::address_space Space> class multi_ptr<void, Space> {
           Accessor)
       : multi_ptr(Accessor.get_pointer()) {}
 
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
+  using ReturnPtr =
+      typename std::conditional<Space == access::address_space::constant_space,
+                                const void *, void *>::type;
+#else
+  using ReturnPtr = void *;
```
Comment (AGindinson) [inline sycl/include/CL/sycl/multi_ptr.hpp]: Maybe `VoidPtr` or `ReturnVoidPtr` instead? (Minor, optional, feel free to ignore)

File: sycl/include/CL/sycl/multi_ptr.hpp [hunk 5]
```diff
@@ -517,6 +548,14 @@ template <typename ElementType, access::address_space Space>
...
   return multi_ptr<ElementType, Space>(pointer);
 }
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
+template <typename ElementType, access::address_space Space,
+          typename = typename std::enable_if<
+              Space == access::address_space::constant_space>::type>
+multi_ptr<ElementType, Space> make_ptr(const ElementType *pointer) {
+  return multi_ptr<ElementType, Space>(pointer);
+}
+#endif
```
Comment (AGindinson) [inline sycl/include/CL/sycl/multi_ptr.hpp]:
```suggestion
#endif // RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR
```

File: sycl/include/CL/sycl/multi_ptr.hpp [hunk 6]
```diff
@@ -517,6 +548,14 @@ template <typename ElementType, access::address_space Space>
...
 #endif
```
Comment (AGindinson) [inline sycl/include/CL/sycl/multi_ptr.hpp]:
```suggestion
#endif // __SYCL_DEVICE_ONLY__
```

**Comment by bader** [comment]
@Fznamznon, please, address comments from @AGindinson.

**Comment by bader** [comment]
@Fznamznon, please, address comments from @AGindinson.

**Comment by romanovvlad** [comment]
@Fznamznon ping

**Comment by Fznamznon** [comment]
@bader , do we need this patch?

**Comment by bader** [comment]
> @bader , do we need this patch?

We definitely need solution for "writing to raw pointers obtained from constant_ptr" problem.
If you have better ideas how to fix this problem, we can investigate them. 
Otherwise, I think this patch is useful.

File: sycl/doc/SYCLPreprocessorMacroses.md (hunk 1 above)
Comment (Fznamznon) [reply sycl/doc/SYCLPreprocessorMacroses.md]: Done.

File: sycl/include/CL/sycl/accessor.hpp (hunk 2 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/accessor.hpp]: Done.

File: sycl/include/CL/sycl/multi_ptr.hpp (hunk 3 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/multi_ptr.hpp]: I've made something like this...

File: sycl/include/CL/sycl/multi_ptr.hpp (hunk 4 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/multi_ptr.hpp]: Not sure about it. This is specialization of multi_ptr class from void type, I think `ReturnPtr ` fits here.

File: sycl/include/CL/sycl/multi_ptr.hpp (hunk 5 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/multi_ptr.hpp]: Done.

File: sycl/include/CL/sycl/multi_ptr.hpp (hunk 6 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/multi_ptr.hpp]: Done.

File: sycl/include/CL/sycl/multi_ptr.hpp (hunk 3 above)
Comment (AGindinson) [reply sycl/include/CL/sycl/multi_ptr.hpp]: Something more C++-conformant, dare I say :)

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Overview
+
+This file describes macros that have effect on SYCL compiler and run-time.
+
+### RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR
+
+Spec assumes that SYCL implementation does address space deduction, but we made
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
The spec assumes that the SYCL implementation does address space deduction. However,
for our implementation, the deduction is performed in the middle end, where it's
```

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+an implementation doing address space deduction in the middle end, where it's
+hard to provide user friendly diagnositcs.
+Due to these problems writing to raw pointers obtained from `constant_ptr` is not
+diagnosed now.
+User can enable diagnostics of writing to such pointers via enabling of
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
The user can enable diagnostics upon writing to such pointers via enabling the
`RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` macro.
```

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+`RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` preprocessor macro.
+Enabling this macro allows `constant_ptr` to use constant pointers as underlying
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
This allows `constant_ptr` to use constant pointers as underlying
```

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+pointer types. So, conversions from `constant_ptr` to raw pointers will return
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
pointer types. Thus, conversions from `constant_ptr` to raw pointers will return
```

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+constant pointers and writing to const pointers will be diagnosed by the
+front-end.
+This behavior is not following SYCL spec since `constant_ptr` conversions to the
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
This behavior is not following the SYCL spec since `constant_ptr` conversions to the
```

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
+# Overview
+
+This file describes macros that have effect on SYCL compiler and run-time.
+
+### RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR
+
+The spec assumes that the SYCL implementation does address space deduction. However,
+for our implementation, the deduction is performed in the middle end, where it's
+an implementation doing address space deduction in the middle end, where it's
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
```

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+hard to provide user friendly diagnositcs.
+Due to these problems writing to raw pointers obtained from `constant_ptr` is not
+diagnosed now.
+The user can enable diagnostics upon writing to such pointers via enabling the
+`RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` macro.
+`RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` preprocessor macro.
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]:
```suggestion
```

File: sycl/include/CL/sycl/multi_ptr.hpp [hunk 14]
```diff
@@ -46,6 +46,13 @@ template <typename ElementType, access::address_space Space> class multi_ptr {
     // TODO An implementation should reject an argument if the deduced
     // address space is not compatible with Space.
   }
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
+  template <access::address_space _Space = Space,
+            typename = typename std::enable_if<
+                _Space == Space &&
```
Comment (rolandschulz) [inline sycl/include/CL/sycl/multi_ptr.hpp]: This line is unnecessary. `_Space == Space` is guaranteed to be true (constructor template arguments can't be explicitly specified).

File: sycl/include/CL/sycl/accessor.hpp [hunk 15]
```diff
@@ -669,9 +678,18 @@ class accessor :
   using AccessorSubscript =
       typename AccessorCommonT::template AccessorSubscript<Dims>;
 
-  using RefType = DataT &;
   using ConcreteASPtrType = typename detail::PtrValueType<DataT, AS>::type *;
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
+  using RefType =
+      typename std::conditional<AS == access::address_space::constant_space,
```
Comment (rolandschulz) [inline sycl/include/CL/sycl/accessor.hpp]: This expression is used at least 6 times. Seems worth to factor out. Suggest change to:
```c++
//somewhere in traits:
#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
template<class AS, class DataT> using const_if_const_AS = typename std::conditional<AS == 
   access::address_space::constant_space, const DataT, DataT>::type;
#else
template<class AS, class DataT> using const_if_const_AS= DataT;
#endif
//and then here:
using RefType = const_if_const_AS<AS, DataT>&;
using PtrType = const_if_const_AS<AS, DataT>*;
```

File: sycl/include/CL/sycl/multi_ptr.hpp [hunk 16]
```diff
@@ -46,6 +46,13 @@ template <typename ElementType, access::address_space Space> class multi_ptr {
     // TODO An implementation should reject an argument if the deduced
     // address space is not compatible with Space.
   }
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
```
Comment (rolandschulz) [inline sycl/include/CL/sycl/multi_ptr.hpp]: Why is this needed at all. Why not change the constructor on line 45 to from `multi_ptr(ElementType*)` to `multi_ptr(pointer_t)`. Then this extra overload isn't needed, correct?

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 17]
```diff
@@ -0,0 +1,22 @@
+# Overview
```
Comment (rolandschulz) [inline sycl/doc/SYCLPreprocessorMacroses.md]: Why is this file in this folder? Shouldn't this go into e.g. doc/extensions/contant_ptr?

Comment (bader) [reply sycl/doc/SYCLPreprocessorMacroses.md]: According to my understanding this not an extension of the functionality defined by the specification, but rather an implementation detail.

**Comment by bader** [comment]
@Fznamznon, please, resolve merge conflict and address comments.

File: sycl/include/CL/sycl/multi_ptr.hpp (hunk 16 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/multi_ptr.hpp]: The spec requires both constructors, the `multi_ptr(pointer_t)` also is defined.

File: sycl/include/CL/sycl/accessor.hpp (hunk 15 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/accessor.hpp]: Sure, done.

File: sycl/include/CL/sycl/multi_ptr.hpp (hunk 14 above)
Comment (Fznamznon) [reply sycl/include/CL/sycl/multi_ptr.hpp]: Removed.

File: sycl/include/CL/sycl/accessor.hpp [hunk 18]
```diff
@@ -682,9 +687,14 @@ class accessor :
   using AccessorSubscript =
       typename AccessorCommonT::template AccessorSubscript<Dims>;
 
-  using RefType = DataT &;
   using ConcreteASPtrType = typename detail::PtrValueType<DataT, AS>::type *;
+#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]: I think we use these definitions even if `RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` is not defined.
```c++
  using RefType = detail::const_if_const_AS<AS, DataT> &;
  using PtrType = detail::const_if_const_AS<AS, DataT> *;
```

Thanks to definition:
```c++
#if defined(RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR)
template <access::address_space AS, class DataT>
using const_if_const_AS =
    typename std::conditional<AS == access::address_space::constant_space,
                              const DataT, DataT>::type;
#else
template <access::address_space AS, class DataT>
using const_if_const_AS = DataT;
#endif
```
Right?
If so, I think there are a few other places where it applies as well.

Comment (Fznamznon) [reply sycl/include/CL/sycl/accessor.hpp]: Sure, done.

**Comment by bader** [comment]
@rolandschulz, @AGindinson, any other comments?

File: sycl/doc/SYCLPreprocessorMacroses.md (hunk 17 above)
Comment (AGindinson) [reply sycl/doc/SYCLPreprocessorMacroses.md]: To keep up with the spirit of "implementation details are DPC++", I'd suggest renaming this into `PreprocessorMacros.md`

File: sycl/doc/SYCLPreprocessorMacroses.md [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+
+### RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR
+
+The spec assumes that the SYCL implementation does address space deduction. However,
+for our implementation, the deduction is performed in the middle end, where it's
+hard to provide user friendly diagnositcs.
+Due to these problems writing to raw pointers obtained from `constant_ptr` is not
+diagnosed now.
+The user can enable diagnostics upon writing to such pointers via enabling the
+`RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR` macro.
```
Comment (AGindinson) [inline sycl/doc/SYCLPreprocessorMacroses.md]: The LIT tests don't seem to be covering this. Is this intentional?

Comment (Fznamznon) [reply sycl/doc/SYCLPreprocessorMacroses.md]: Diagnostics will point to SYCL headers, I don't know the way how to validate diagnostics pointing not to test file. If you know, we can probably add it. But, I'm not sure that we have to validate standard c++ diagnostics because this patch just re-uses them.

File: sycl/doc/SYCLPreprocessorMacroses.md (hunk 17 above)
Comment (Fznamznon) [reply sycl/doc/SYCLPreprocessorMacroses.md]: Sure.

File: sycl/doc/SYCLPreprocessorMacroses.md (hunk 19 above)
Comment (AGindinson) [reply sycl/doc/SYCLPreprocessorMacroses.md]: Thanks, got it.
